Release notes v0.47.0 (#3360)
grafana/k6: release notes/v0.47.0.md

k6 `v0.47.0` is here 🎉! This release includes:

> ## Breaking changes

Reviewer: I agree with @heitortsergent on [the previous comment](https://github.com/grafana/k6/pull/3360#discussion_r1343132051) and want to extend it with the suggestion of the whole section renaming. 

```suggestion
## Deprecations
```

The thing is that none of the mentioned things doesn't break k6 right now, but we will do this in the future. But seems reasonable to reserve the `breaking changes` for the fundamental breaking changes to keep it urgency.

The k6 gRPC modules (`k6/net/grpc` and `k6/experimental/grpc`) now support handling binary metadata that uses the `-bin` postfix, according to the gRPC specification.

```js
let resp = client.invoke("grpc.testing.TestService/EmptyCall", {}, { metadata: { "X-Load-Tester-bin": new Uint8Array([2, 200]) } })
```

Thanks to @sapphire-janrain for the contribution!

### Add gRPC's reflection metadata support [#3343](https://github.com/grafana/k6/pull/3343), [xk6-grpc#46](https://github.com/grafana/xk6-grpc/pull/46)

The k6 gRPC modules (`k6/net/grpc` and `k6/experimental/grpc`) now support adding metadata to reflection requests by using a new connection parameter [`reflectMetadata`](https://k6.io/docs/javascript-api/k6-net-grpc/client/client-connect/#connectparams).

### Higher precision for Trend metrics in Cloud [#3302](https://github.com/grafana/k6/pull/3302)

k6 Cloud is now able to store and visualize Trend metrics up to 3 digits of precision for decimal numbers.

### Docker images for ARM64 architecture [#3320](https://github.com/grafana/k6/pull/3320)

> k6 now builds and pushes dedicated Docker images for ARM64. Check k6's [tags page](https://hub.docker.com/r/grafana/k6/tags) on DockerHub for details. 

Reviewer: It's not really k6, so it's either neutral we or maybe k6's CI.

```suggestion
We started to build and push dedicated Docker images for ARM64. Check k6's [tags page](https://hub.docker.com/r/grafana/k6/tags) on DockerHub for details. 
```

Author: Applied,  not exactly the same but very similar.

```js
const context = browser.newContext();
context.addCookies([
    {name: 'foo', value: 'bar', url: 'https://test.k6.io'},
    {name: 'baz', value: 'qux', url: 'https://grafana.com'},
]);

const cookies = context.cookies('https://test.k6.io');
console.log(cookies.length);         // 1
console.log(cookies[0].name);        // foo
console.log(cookies[0].value);       // bar

context.clearCookies();
console.log(context.cookies.length); // 0
```

### Add support for browser module's `page.on('console')` [browser#1006](https://github.com/grafana/xk6-browser/pull/1006)

> Allows users to register a handler to be executed every time the `console` API methods are called from within the page's JavaScript context. The arguments passed into the handler are defined by the [ConsoleMessage](https://k6.io/docs/javascript-api/k6-experimental/browser/consolemessage/) class.

Reviewer: ```suggestion
This support allows users to register a handler to be executed every time the `console` API methods are called from within the page's JavaScript context. The arguments passed into the handler are defined by the [ConsoleMessage](https://k6.io/docs/javascript-api/k6-experimental/browser/consolemessage/) class.
```

Reviewer 2: I think the original message is good enough, thanks.

```js
page.on('console', msg => {
    check(msg, {
        'assertConsoleMessageType': msg => msg.type() == 'log',
        'assertConsoleMessageText': msg => msg.text() == 'this is a console.log message 42',
        'assertConsoleMessageArgs0': msg => msg.args()[0].jsonValue() == 'this is a console.log message',
        'assertConsoleMessageArgs1': msg => msg.args()[1].jsonValue() == 42,
    });
});

page.evaluate(() => console.log('this is a console.log message', 42));
```

> ### UX improvements and enhancements

Reviewer 3: ```suggestion
## UX improvements and enhancements
```

nit: should we switch this to a h2?

Author: Done, we also updated the template as it was generated from it

- [#3338](https://github.com/grafana/k6/pull/3338), [xk6-grpc#48](https://github.com/grafana/xk6-grpc/pull/48) Adds support for the gRPC reflection protocol v1.

> - [#3290](https://github.com/grafana/k6/pull/3290) Adds error logging when executing `setup` and `teardown` via REST API.

Reviewer: ```suggestion
- [#3290](https://github.com/grafana/k6/pull/3290) Adds error logging when executing `setup` and `teardown` via REST API. Thanks, @kmtym1998, for the contribution!
```

- [#3340](https://github.com/grafana/k6/pull/3340) Updates k6 ``*-with-browser`` Docker images to automatically set the `no-sandbox` environment variable.
- [#3335](https://github.com/grafana/k6/pull/3335) The character limit for metric names increased from 63 to 128 after the OpenTelemetry update. k6 will return an error starting on the next release (v0.48.0) if users hit the limit.
- [browser#1007](https://github.com/grafana/xk6-browser/pull/1007) Adds a `k6` object (`window.k6 = {};`) to help identify k6 browser module tests.
- [browser#1022](https://github.com/grafana/xk6-browser/pull/1022) Refactors the `check` in `examples/fillform.js` so that it matches the type definitions and documentation for `check`.

## Bug fixes

- [xk6-grpc#47](https://github.com/grafana/xk6-grpc/pull/47) Fixes the premature closing of a gRPC stream when a stream's client has finished sending. Thanks to @thiagodpf for reporting!
- [#3344](https://github.com/grafana/k6/pull/3344), [xk6-grpc#49](https://github.com/grafana/xk6-grpc/pull/49) Adds support for Google's protobuf wrappers. Thanks to @zibul444 for reporting!

> - [#3346](https://github.com/grafana/k6/pull/3346) Fixes a race condition when `http.batch` is used with a binary response.

Reviewer: I'm wondering if this should be in a bug fixes section. If I remember correctly, the bug was caused by the goja dependency update, spotted in the tests and fixed. So maybe it's more about maintenance or improvements or even "squash" with the line about the goja update. 

@mstoykov, what do you think?

Author: Let's squash then, I'm going to apply it :+1:

Author: I just dropped it as it wasn't an issue in the previous version so it doesn't make sense to mention the fix.

- [#3308](https://github.com/grafana/k6/pull/3308) Updates `goja` version, and fixes a compiler bug when a class is declared in a function with an argument.

> - [browser#1040](https://github.com/grafana/xk6-browser/pull/1040) Fixes the `context.clearCookies()` method to clear all cookies from the current browser context.

Reviewer: These two:

```
- [browser#1040](https://github.com/grafana/xk6-browser/pull/1040) Fixes the `context.clearCookies()` method to clear all cookies from the current browser context.
- [browser#1031](https://github.com/grafana/xk6-browser/pull/1031) Fixes the `expires` field while adding cookies using the `context.addCookie()` method.
```

Were already mentioned in the `Improved the browser module's cookie API`.

Maybe just link these from there to avoid deduplication?

Author: @ankur22 WDYT?

Reviewer 2: Resolved in 17b1d74da7eb74b4bf59aeab7b660c68deb83a68. Could you double check that the rewording makes sense or not?

Reviewer: TBH, I'm not sure that I'm the best person for checking wording :sweat_smile: The only thing that I remember is that we should avoid using passive :shrug: 

But we could ask @heitortsergent  to help us with checking the rewording :relaxed:

Reviewer 3: The changes look good to me. 👍

- [browser#1031](https://github.com/grafana/xk6-browser/pull/1031) Fixes the `expires` field while adding cookies using the `context.addCookie()` method.
- [browser#1039](https://github.com/grafana/xk6-browser/pull/1039) Fixes `goja` conversions while adding and retrieving cookies.
- [browser#1038](https://github.com/grafana/xk6-browser/pull/1038) Fixes read/write data race for edge case with remote browsers.

> - [browser#1035](https://github.com/grafana/xk6-browser/pull/1035) Refactors `int64` timeout to `time.Duration`, to help avoid confusion as to whether a timeout is in milliseconds or seconds.

Reviewer: TBH It doesn't sound like a bug fix, or at least not user-faced, how about moving this to the maintenance and internal improvements? :thinking:

Author: @ankur22 do you confirm?

Reviewer 2: Yeah, sure, i guess that makes sense. Do you want me to make the change?

Reviewer 2: Resolved in bc3eadba973acca7bb531501d5be0983b86b2ad8

- [browser#1034](https://github.com/grafana/xk6-browser/pull/1034) Fixes `page.reload` & `page.setContent` to use the default navigation timeout over the default timeout.
- [browser#1033](https://github.com/grafana/xk6-browser/pull/1033) Fixes the `page` timeouts so it is actually used after being set.

## Maintenance and internal improvements

- [#3342](https://github.com/grafana/k6/pull/3342) Updates xk6-grpc to the latest version. This change brings all the latest fixes and improvements to the experimental gRPC module.
- [#3271](https://github.com/grafana/k6/pull/3271),[#3272](https://github.com/grafana/k6/pull/3272) Updates the golangci version and adds the `interfacebloat` linter.
- [#3279](https://github.com/grafana/k6/pull/3279) Fixes the CI not publishing the SBOM file on a new release.

> - [#3283](https://github.com/grafana/k6/pull/3283) Updates the Go version to 1.21.

Reviewer: I remember the @mstoykov [was pointing out](https://github.com/grafana/k6/pull/3283#issuecomment-1681742790) that is worth explicitly mentioning this happens in CI. 

Otherwise, it could be interpreted as switching the versions in the k6, like starting to use some functionality from go1.21, which will require a go1.21.

```suggestion
- [#3283](https://github.com/grafana/k6/pull/3283) Updates the Go version in k6's CI to 1.21.
```

Author: For me the important part is that k6 is built using `go1.21`, so I would clarify a bit more. What about the following?

```suggestion
- [#3283](https://github.com/grafana/k6/pull/3283) Updates the Go version in k6's CI used to build the binaries. The minimum supported version for extensions continues to be `go1.19`.
```

Reviewer: :+1:  for the `build the binaries...` and soft :-1: for the minimum supported version since, in my opinion, the sentence doesn't bring any essential details, but extends the cognitive load of the sentence

- [#3311](https://github.com/grafana/k6/pull/3311) Updates the `alpine` image version that is used as the base of the k6 Docker image.
- [browser#1043](https://github.com/grafana/xk6-browser/pull/1043), [browser#1021](https://github.com/grafana/xk6-browser/pull/1021), [browser#1019](https://github.com/grafana/xk6-browser/pull/1019), [browser#1014](https://github.com/grafana/xk6-browser/pull/1014) Fixes xk6-browser tests.
- [browser#1000](https://github.com/grafana/xk6-browser/pull/1000), [browser#1024](https://github.com/grafana/xk6-browser/pull/1024) Refines xk6-browser issue and PR templates.
- [browser#1003](https://github.com/grafana/xk6-browser/pull/1003), [browser#1009](https://github.com/grafana/xk6-browser/pull/1009), [browser#1010](https://github.com/grafana/xk6-browser/pull/1010) Internal changes to xk6-browser.
- [browser#997](https://github.com/grafana/xk6-browser/pull/997) Updates xk6-browser readme.
- [browser#962](https://github.com/grafana/xk6-browser/pull/962) CI fixes.

## _Optional_ Roadmap

_Discussion of future plans_

> 

Reviewer 4: Shall we mention har converter will be dropped as well as https://github.com/grafana/k6/issues/3065 and https://github.com/grafana/k6/issues/3287 ?

k6 `v0.47.0` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`
- `<Summary of new features>` (_one or multiple bullets_)

## Breaking changes

> - [#3347](https://github.com/grafana/k6/pull/3347) `statsd` output is now deprecated. Check out [#2982](https://github.com/grafana/k6/issues/2982) for the future plans.

Author: @heitortsergent Are you going to address it?

> - [#3347](https://github.com/grafana/k6/pull/3347) `statsd` output is now deprecated. Check out [#2982](https://github.com/grafana/k6/issues/2982) for the future plans.

Reviewer 3: Yeah, I'll open a PR for it. :)

> - [#3335](https://github.com/grafana/k6/pull/3335) Metric names will be limited to 128 characters following the OpenTelemetry update instead of the previous limit of 63 characters. k6 will return an error from the next release (v0.48.0) if the limit is hit.

Author: Not really, good point

> - [#3335](https://github.com/grafana/k6/pull/3335) Metric names will be limited to 128 characters following the OpenTelemetry update instead of the previous limit of 63 characters. k6 will return an error from the next release (v0.48.0) if the limit is hit.

Author: @mstoykov I would like to move this to `UX Improvements - Enhancements` section. Is it fine for you, as you set it originally as a breaking change?

> - [#3335](https://github.com/grafana/k6/pull/3335) Metric names will be limited to 128 characters following the OpenTelemetry update instead of the previous limit of 63 characters. k6 will return an error from the next release (v0.48.0) if the limit is hit.

Reviewer 4: Yeah - let's do that then

## New features

### Add gRPC's binary metadata support [#3234](https://github.com/grafana/k6/pull/3234), [xk6-grpc#46](https://github.com/grafana/xk6-grpc/pull/46)

By specification gRPC metadata with the `-bin` postfix should be treated as binary data. These PRs add the support of such metadata to the `k6/net/grpc` and `k6/experimental/grpc` modules.

```js
let resp = client.invoke("grpc.testing.TestService/EmptyCall", {}, { metadata: { "X-Load-Tester-bin": new Uint8Array([2, 200]) } })
```

Thanks, @sapphire-janrain, for contribution!

### Add the reflection metadata [#3343](https://github.com/grafana/k6/pull/3343), [xk6-grpc#46](https://github.com/grafana/xk6-grpc/pull/46)

In some workflows the reflection call should also include some metadata. This PR adds [a new connection parameter `reflectMetadata`](https://k6.io/docs/javascript-api/k6-net-grpc/client/client-connect/#connectparams) that allows to specify the metadata to be sent with the reflection call.

### Higher precision for Trend metrics in Cloud [#3302](https://github.com/grafana/k6/pull/3302)

> k6 Cloud is now able to store and visualize Trend metrics up to 3 digits of precision for decimal numbers.

Author: It's for both `app.k6.io` and GCk6. Now, I'm not sure what we should use. Grafana Cloud k6 is the present and the future so probably your suggestion makes sense. WDYT?

> k6 Cloud is now able to store and visualize Trend metrics up to 3 digits of precision for decimal numbers.

Reviewer 3: Yeah, let's stick with "Grafana Cloud k6" going forward, unless it's anything specific to just app.k6.io.

> ### Docker images for Arm64 architecture [#3320](https://github.com/grafana/k6/pull/3320)

Author: The architecture will be automatically detected by Docker so in most cases it works out of the box. If they would like to force then they have to explicit it. But in this case, I expect they know how to do that.

```js
page.on('console', msg => {
    check(msg, {
        'assertConsoleMessageType': msg => msg.type() == 'log',
        'assertConsoleMessageText': msg => msg.text() == 'this is a console.log message 42',
        'assertConsoleMessageArgs0': msg => msg.args()[0].jsonValue() == 'this is a console.log message',
        'assertConsoleMessageArgs1': msg => msg.args()[1].jsonValue() == 42,
    });
});

page.evaluate(() => console.log('this is a console.log message', 42));
```

### UX improvements and enhancements

- [#3338](https://github.com/grafana/k6/pull/3338), [xk6-grpc#48](https://github.com/grafana/xk6-grpc/pull/48). Adds a support of the v1 of the gRPC reflection protocol.
- [#3290](https://github.com/grafana/k6/pull/3290) Log errors when executing `setup` and `teardown` via REST API.
- [#3327](https://github.com/grafana/k6/pull/3327) `k6 version` now prints the commit identifier for the build.
- [#3340](https://github.com/grafana/k6/pull/3340) The k6 ``*-with-browser`` Docker images set now the `no-sandbox` environment variable automatically.

> - [browser#1007](https://github.com/grafana/xk6-browser/pull/1007) Inject `window.k6 = {};` to help identify k6 browser module tests.

Author: @grafana/k6-browser let me know which is the version that you prefer

> - [browser#1007](https://github.com/grafana/xk6-browser/pull/1007) Inject `window.k6 = {};` to help identify k6 browser module tests.

Reviewer 2: @codebien I like @heitortsergent suggestion 👍

- [browser#1022](https://github.com/grafana/xk6-browser/pull/1022) Refactor the `check` in `examples/fillform.js` so that it matches the type definitions and documentation for `check`.

## Bug fixes

- [xk6-grpc#47](https://github.com/grafana/xk6-grpc/pull/47) Fix the premature closing of the stream when a stream's client has finished sending. Thanks, @thiagodpf, for reporting!
- [#3344](https://github.com/grafana/k6/pull/3344), [xk6-grpc#49](https://github.com/grafana/xk6-grpc/pull/49) adds (fixes) the support of google protobuf wrappers. Thanks, @zibul444, for reporting!
- [#3346](https://github.com/grafana/k6/pull/3346) A data race when `http.batch` it is used with binary response has been fixed.
- [#3308](https://github.com/grafana/k6/pull/3308) `goja` version has been updated and it fixes a compiler bug when class is declared in a function with argument.

> - [browser#1040](https://github.com/grafana/xk6-browser/pull/1040) fix the `context.clearCookies()` method to clear all cookies from the current browser context.
> - [browser#1031](https://github.com/grafana/xk6-browser/pull/1031) fix the `expires` field while adding cookies using the `context.addCookie()` method.
> - [browser#1039](https://github.com/grafana/xk6-browser/pull/1039) fix Goja conversions while adding and retrieving cookies.
> - [browser#1038](https://github.com/grafana/xk6-browser/pull/1038) Fix read/write data race for edge case with remote browsers.
> - [browser#1035](https://github.com/grafana/xk6-browser/pull/1035) Refactor `int64` timeout to `time.Duration` which ensures no confusion as to whether a timeout is in milliseconds or seconds.
> - [browser#1034](https://github.com/grafana/xk6-browser/pull/1034) Fix `page.reload` & `page.setContent` to use the default navigation timeout over the default timeout.
> - [browser#1033](https://github.com/grafana/xk6-browser/pull/1033) Fix the `page` timeouts so it is actually used after being set.

Reviewer 3: ```suggestion
- [browser#1040](https://github.com/grafana/xk6-browser/pull/1040) Fixes the `context.clearCookies()` method to clear all cookies from the current browser context.
- [browser#1031](https://github.com/grafana/xk6-browser/pull/1031) Fixes the `expires` field while adding cookies using the `context.addCookie()` method.
- [browser#1039](https://github.com/grafana/xk6-browser/pull/1039) Fixes `goja` conversions while adding and retrieving cookies.
- [browser#1038](https://github.com/grafana/xk6-browser/pull/1038) Fixes read/write data race for edge case with remote browsers.
- [browser#1035](https://github.com/grafana/xk6-browser/pull/1035) Refactors `int64` timeout to `time.Duration`, to help avoid confusion as to whether a timeout is in milliseconds or seconds.
- [browser#1034](https://github.com/grafana/xk6-browser/pull/1034) Fixes `page.reload` & `page.setContent` to use the default navigation timeout over the default timeout.
- [browser#1033](https://github.com/grafana/xk6-browser/pull/1033) Fixes the `page` timeouts so it is actually used after being set.
```

## Maintenance and internal improvements

- [#3342](https://github.com/grafana/k6/pull/3342) Updates xk6-grpc to the latest version. This change brings all the latest fixes and improvements to the experimental gRPC module.

> - [#3271](https://github.com/grafana/k6/pull/3271),[#3272](https://github.com/grafana/k6/pull/3272) Updated the golangci version and added the `interfacebloat` linter.
> - [#3279](https://github.com/grafana/k6/pull/3279) Fixed the CI not publishing the SBOM file on new a release.
> - [#3283](https://github.com/grafana/k6/pull/3283) Updated the Go version to 1.21.
> - [#3351](https://github.com/grafana/k6/pull/3351) Updated the version of Prometheus remote write output. Check the specific [release notes](https://github.com/grafana/xk6-output-prometheus-remote/releases/tag/v0.3.0).
> - [#3341](https://github.com/grafana/k6/pull/3341), [#3339](https://github.com/grafana/k6/pull/3339) Updated `goja` that includes Runtime initialization speed-up and a fix for source indexes. 
> - [#3311](https://github.com/grafana/k6/pull/3311) Updated the `alpine` image version that is used as the base of the k6 Docker image.
> - [browser#1043](https://github.com/grafana/xk6-browser/pull/1043), [browser#1021](https://github.com/grafana/xk6-browser/pull/1021), [browser#1019](https://github.com/grafana/xk6-browser/pull/1019), [browser#1014](https://github.com/grafana/xk6-browser/pull/1014) Fix tests.
> - [browser#1000](https://github.com/grafana/xk6-browser/pull/1000), [browser#1024](https://github.com/grafana/xk6-browser/pull/1024) Refine issue and PR templates.
> - [browser#1003](https://github.com/grafana/xk6-browser/pull/1003), [browser#1009](https://github.com/grafana/xk6-browser/pull/1009), [browser#1010](https://github.com/grafana/xk6-browser/pull/1010) Internal changes.
> - [browser#997](https://github.com/grafana/xk6-browser/pull/997) Update readme.
> - [browser#962](https://github.com/grafana/xk6-browser/pull/962) CI fixes.

Reviewer 3: ```suggestion
- [#3271](https://github.com/grafana/k6/pull/3271),[#3272](https://github.com/grafana/k6/pull/3272) Updates the golangci version and adds the `interfacebloat` linter.
- [#3279](https://github.com/grafana/k6/pull/3279) Fixes the CI not publishing the SBOM file on a new release.
- [#3283](https://github.com/grafana/k6/pull/3283) Updates the Go version to 1.21.
- [#3351](https://github.com/grafana/k6/pull/3351) Updates the version of Prometheus remote write output. Refer to the specific [release notes](https://github.com/grafana/xk6-output-prometheus-remote/releases/tag/v0.3.0) for more details.
- [#3341](https://github.com/grafana/k6/pull/3341), [#3339](https://github.com/grafana/k6/pull/3339) Updates `goja`, includes runtime initialization speed-up and a fix for source indexes. 
- [#3311](https://github.com/grafana/k6/pull/3311) Updates the `alpine` image version that is used as the base of the k6 Docker image.
- [browser#1043](https://github.com/grafana/xk6-browser/pull/1043), [browser#1021](https://github.com/grafana/xk6-browser/pull/1021), [browser#1019](https://github.com/grafana/xk6-browser/pull/1019), [browser#1014](https://github.com/grafana/xk6-browser/pull/1014) Fixes xk6-browser tests.
- [browser#1000](https://github.com/grafana/xk6-browser/pull/1000), [browser#1024](https://github.com/grafana/xk6-browser/pull/1024) Refines xk6-browser issue and PR templates.
- [browser#1003](https://github.com/grafana/xk6-browser/pull/1003), [browser#1009](https://github.com/grafana/xk6-browser/pull/1009), [browser#1010](https://github.com/grafana/xk6-browser/pull/1010) Internal changes to xk6-browser.
- [browser#997](https://github.com/grafana/xk6-browser/pull/997) Updates xk6-browser readme.
- [browser#962](https://github.com/grafana/xk6-browser/pull/962) CI fixes.
```

- [browser#1035](https://github.com/grafana/xk6-browser/pull/1035) Refactors `int64` timeout to `time.Duration`, to help avoid confusion as to whether a timeout is in milliseconds or seconds.
- [browser#1034](https://github.com/grafana/xk6-browser/pull/1034) Fixes `page.reload` & `page.setContent` to use the default navigation timeout over the default timeout.
- [browser#1033](https://github.com/grafana/xk6-browser/pull/1033) Fixes the `page` timeouts so it is actually used after being set.

## Maintenance and internal improvements

- [#3342](https://github.com/grafana/k6/pull/3342) Updates xk6-grpc to the latest version. This change brings all the latest fixes and improvements to the experimental gRPC module.
- [#3271](https://github.com/grafana/k6/pull/3271),[#3272](https://github.com/grafana/k6/pull/3272) Updates the golangci version and adds the `interfacebloat` linter.
- [#3279](https://github.com/grafana/k6/pull/3279) Fixes the CI not publishing the SBOM file on a new release.
- [#3283](https://github.com/grafana/k6/pull/3283) Updates the Go version in k6's CI used to build the binaries.

> - [#3351](https://github.com/grafana/k6/pull/3351) Updates the version of Prometheus remote write output. Refer to the specific [release notes](https://github.com/grafana/xk6-output-prometheus-remote/releases/tag/v0.3.0) for more details.

Reviewer: I wonder if it makes sense to make this a feature to give more insights and make it more transparent/valuable for the users?

Maybe you could mention which new ENV variables bring this to the experimental k6 and how to use them? :thinking:

Author: Done

For users, it will mean better JavaScript support as this change will automatically get [object spread](https://github.com/grafana/k6/issues/824) working and likely faster startup for *big* scripts. In the future, this also means JavaScript compatibility will be easier to add, since it'll only need to be supported in the JavaScript VM we use - [goja](https://github.com/dop251/goja).

There's a risk that some k6 tests using both CommonJS and ECMAScript modules syntax will change in behavior. In practice, using both should never really be done as they're not compatible.

Because there are risks involved, we are going to do this very carefully and only once we have done a lot of tests. That means this is currently planned for v0.49.0, but with the majority of the remaining work done in the v0.48.0 cycle.

This way, we will also be able to have most of our users test these changes from our `master` branch, using Docker images, for example.

### Future breaking changes

> It's relevant to stretch that in the next release several breaking changes will be applied as final stage of their deprecation period:

Reviewer 3: ```suggestion
There are several changes in the next release that are entering the final stage of their deprecation period. That means the next release will include the following breaking changes:
```

### Higher precision for Trend metrics in Grafana Cloud k6 [#3302](https://github.com/grafana/k6/pull/3302)

Grafana Cloud k6 is now able to store and visualize Trend metrics up to 3 digits of precision for decimal numbers.

### Docker support for browser-based tests [#3199](https://github.com/grafana/k6/pull/3199)

k6 is now publishig Docker images that include Chromium web browser. This allows k6 users to run tests that use [Browser API](https://k6.io/docs/javascript-api/k6-experimental/browser/) without having to install Chrome first. Check the ["A note on running browser tests" section](https://hub.docker.com/r/grafana/k6) of the Overview page on DockerHub for details.

### Docker images for ARM64 architecture [#3320](https://github.com/grafana/k6/pull/3320)

The k6's release process now builds and pushes dedicated Docker images for ARM64. Check k6's [tags page](https://hub.docker.com/r/grafana/k6/tags) on DockerHub for details.

> ### New authentication methods and HTTP headers API for Prometheus remote write output 

Reviewer 3: ```suggestion
### New authentication methods and HTTP headers API for Prometheus remote write output [xk6-output-prometheus-remote#143](https://github.com/grafana/xk6-output-prometheus-remote/pull/143), [xk6-output-prometheus-remote#145](https://github.com/grafana/xk6-output-prometheus-remote/pull/145), [xk6-output-prometheus-remote#147](https://github.com/grafana/xk6-output-prometheus-remote/pull/147)
```

These seem to be the relevant PRs, but please feel free to edit. 🙏

> The experimental Prometheus remote write output gets the support for two new authentication methods: Bearer token and TLS certificates. Check out the [documentation](https://k6.io/docs/results-output/real-time/prometheus-remote-write/#options) to know how to define them using the new environement variables.

Author: Set as a reminder, we need to add them to the documentation

> The experimental Prometheus remote write output gets the support for two new authentication methods: Bearer token and TLS certificates. Check out the [documentation](https://k6.io/docs/results-output/real-time/prometheus-remote-write/#options) to know how to define them using the new environement variables.

Reviewer 3: ```suggestion
The experimental Prometheus remote write output now supports two new authentication methods: Bearer token and TLS certificates. Check out the [documentation](https://k6.io/docs/results-output/real-time/prometheus-remote-write/#options) to learn more about how to define them using the new environment variables.
```

- [browser#1043](https://github.com/grafana/xk6-browser/pull/1043), [browser#1021](https://github.com/grafana/xk6-browser/pull/1021), [browser#1019](https://github.com/grafana/xk6-browser/pull/1019), [browser#1014](https://github.com/grafana/xk6-browser/pull/1014) Fixes xk6-browser tests.
- [browser#1000](https://github.com/grafana/xk6-browser/pull/1000), [browser#1024](https://github.com/grafana/xk6-browser/pull/1024) Refines xk6-browser issue and PR templates.
- [browser#1003](https://github.com/grafana/xk6-browser/pull/1003), [browser#1009](https://github.com/grafana/xk6-browser/pull/1009), [browser#1010](https://github.com/grafana/xk6-browser/pull/1010) Internal changes to xk6-browser.
- [browser#997](https://github.com/grafana/xk6-browser/pull/997) Updates xk6-browser readme.
- [browser#962](https://github.com/grafana/xk6-browser/pull/962) CI fixes.

## Roadmap

### Native ECMAScript Modules support

Work on this [epic issue](https://github.com/grafana/k6/issues/3265) has been picked up and there is some progress in the underlying implementation.

> One of the main internal changes will be dropping of our internal [Babel](https://babeljs.io) that is currently used to transpile [ESM](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) code to [CommonJS](https://en.wikipedia.org/wiki/CommonJS) one.

Reviewer 3: ```suggestion
One of the main internal changes will be dropping [Babel](https://babeljs.io), which is currently used to transpile [ESM](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) code to [CommonJS](https://en.wikipedia.org/wiki/CommonJS).
```

> For users it will mean better JavaScript support as this for example will automatically get [object spread](https://github.com/grafana/k6/issues/824) working and likely faster startup for *big* scripts. In the future this also means easier adding of JavaScript compatibility as now it needs to be supported only in the JavaScript VM we use - [goja](https://github.com/dop251/goja)

Reviewer 3: ```suggestion
For users, it will mean better JavaScript support as this change will automatically get [object spread](https://github.com/grafana/k6/issues/824) working and likely faster startup for *big* scripts. In the future, this also means JavaScript compatibility will be easier to add, since it'll only need to be supported in the JavaScript VM we use - [goja](https://github.com/dop251/goja).
```

I tried to clarify things here, but please let me know if this still makes sense from a technical perspective.

> There is risk though that some k6 tests using both CommonJS and ECMAScript Modules syntax will change in behavior. This in practice should never really be done as both are not compatible.

Reviewer 3: ```suggestion
There's a risk that some k6 tests using both CommonJS and ECMAScript modules syntax will change in behavior. In practice, using both should never really be done as they're not compatible.
```

> Due to this we are going to be dropping it very carefully and likely only once we have done a lot of tests. This currently means that it is planned for v0.49.0, but with the majority of the remaining work done in v0.48.0 cycle.
> 
> This way as well we will have the most time of users being able to test from our `master` branch, using docker images for example.

Reviewer 3: ```suggestion
Because there are risks involved, we are going to do this very carefully and only once we have done a lot of tests. That means this is currently planned for v0.49.0, but with the majority of the remaining work done in the v0.48.0 cycle.

This way, we will also be able to have most of our users test these changes from our `master` branch, using Docker images, for example.
```

### Future breaking changes

It's relevant to stretch that in the next release several breaking changes will be applied as final stage of their deprecation period:

> - [#3065](https://github.com/grafana/k6/issues/3065) Limit the length of metric names to 128 characters and the set of the allowed symbols.

Reviewer 4: Technically 128 characters is a limti now as well :thinking:

Reviewer 3: Maybe we can be more verbose here and explain what happened? Something like:

"Metric names will be limited to 128 characters and a set of allowed symbols. This was going to be changed to a 63 character limit to ensure compatibility with Otel standards, but since then, the Otel limit has been updated to 255 characters. You can find more details about the changes in the linked issue."

Reviewer 4: I think that is too much especially as we have had 1 report at this being a problem over the months taht we are doing thsi change. And this is already explained in the docs. 

So let's just keep it like this.

- [#3365](https://github.com/grafana/k6/pull/3365) The `k6 converter` command will be removed.

> - [#3350](https://github.com/grafana/k6/pull/3350) The `headears` param for the gRPC module will be not anymore accepted.

Reviewer 4: ```suggestion
- [#3350](https://github.com/grafana/k6/pull/3350) The `headers` param for the gRPC module will be not anymore accepted.
```

k6 `v0.47.0` is here 🎉! This release includes:

- (_optional_) `<highlight of breaking changes>`

> - `<Summary of new features>` (_one or multiple bullets_)

Author: This time the list is already compact so probably we can omit them as it will be probably the same as the feature list. WDYT? If you want to keep then make a suggestion, please.

## Maintenance and internal improvements

- [#3342](https://github.com/grafana/k6/pull/3342) Updates xk6-grpc to the latest version. This change brings all the latest fixes and improvements to the experimental gRPC module.
- [#3271](https://github.com/grafana/k6/pull/3271),[#3272](https://github.com/grafana/k6/pull/3272) Updated the golangci version and added the `interfacebloat` linter.
- [#3279](https://github.com/grafana/k6/pull/3279) Fixed the CI not publishing the SBOM file on new a release.
- [#3283](https://github.com/grafana/k6/pull/3283) Updated the Go version to 1.21.
- [#3351](https://github.com/grafana/k6/pull/3351) Updated the version of Prometheus remote write output. Check the specific [release notes](https://github.com/grafana/xk6-output-prometheus-remote/releases/tag/v0.3.0).
- [#3341](https://github.com/grafana/k6/pull/3341), [#3339](https://github.com/grafana/k6/pull/3339) Updated `goja` that includes Runtime initialization speed-up and a fix for source indexes. 
- [#3311](https://github.com/grafana/k6/pull/3311) Updated the `alpine` image version that is used as the base of the k6 Docker image.

## _Optional_ Roadmap

> _Discussion of future plans_

Author: @mstoykov do you want to include the ECMAScript Modules here?

Author: @mstoykov can you make a suggestion for it, please? You are probably the best for doing it.

Reviewer 4: ```suggestion
### Native Modules Support

Work on this [epic issue](https://github.com/grafana/k6/issues/3265) has been picked up and there is some progress in the underlying implementation. 

One of the main internal changes will be dropping of our internal [babel](https://babeljs.io/) that is currently used to transpile [ESM](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) code to [commonjs]()https://en.wikipedia.org/wiki/CommonJS) one.
 
For users that will mean better JS support as this for example will automatically get [object spread](https://github.com/grafana/k6/issues/824) working and likely faster startup for *big* scripts. In the future this also means easier adding of JS compatibility as now it needs to be supported only in the JS VM we use - [goja](https://github.com/dop251/goja)

There is risk though that some k6 tests using both commonjs and ESM syntax will change in behavior. This in practice should never really be done as both are not compatible.

Due to this we are going to be dropping it very carefully and likely only once we have done a lot of tests. This currently means that it is planned for v0.49.0, but with the majority of the remaining work done in v0.48.0 cycle.

This way as well we will have the most time of users being able to test from our `master` branch, using docker images for example.
```
I tried to give some amount of info so it is less "we are doing this - figure it out", but also tried not to list everything.
